### PR TITLE
Add better casting to models

### PR DESCRIPTION
**Pull request by KieranFYI**
## Contribution type

Bug Fix

## Description of change

When viewing builds I noticed that clicking the commit link didn't open the correct commit, I also noticed this was happening for any file references/line references, this is due to casting in the Model auto converting anything which ended with _id to int but commit_id is expected to be a string. I realise the Model had no proper support for casting, which this branch rectifies.

While adding support for casting I also fixed all the calls to include the correct types and return types and added two new function (getData, setData) which handle casting/storing of the data and marking a column as modified. I've also modified Store to reduce code and cleanup some functions.

Running unit tests on my local install I then noticed that if a phpunit failed the whole worker process would quit due to line parsing. Unsure if this is related to the changes i've made, but it's now been fixed.

Checking the functionality of the site after all the changes, I was unable to logout but this is now fixed.

Sorry for such a huge PR.

### CODE REVIEW

**Comment by KieranFYI** [comment]
I'm currently writing tests to improve the code coverage of models. Please hold off reviewing for now.

**Comment by corpsee** [comment]
@KieranFYI Thank you! I'll do a PR review soon, but I need time to do it. I made a quick fix about `commit_id` field for now: [09e4607](https://github.com/php-censor/php-censor/commit/09e46072582935aa203c066ecde3ff6f2bd8def3)

**Comment by corpsee** [comment]
In addition I think we should rename `commit_id` to `commit_hash` in the future. This would be semantically more correct.

File: src/Model.php [hunk 1]
```diff
@@ -15,31 +18,49 @@ class Model
 {
     protected array $data = [];
 
-    protected array $modified = [];
+    protected array $casts = [];
```
Comment (corpsee) [inline src/Model.php:21]: I think `$dataTypes` would be a better name for `$casts`.

File: src/Model.php [hunk 2]
```diff
@@ -50,10 +71,67 @@ public function getModified(): array
...
     {
-        $this->modified[$column] = $column;
+        if ($value === null) {
+            return null;
+        }
 
-        return true;
+        switch ($type) {
+            case 'int':
+            case 'integer':
```
Comment (corpsee) [inline src/Model.php:85]: You can make types constants.

File: src/Model.php [hunk 3]
```diff
@@ -15,31 +18,49 @@ class Model
...
 
         $this->storeRegistry = $storeRegistry;
     }
 
+    public function getCast(string $column): string
+    {
+        return $this->casts[$column] ?? 'string';
+    }
+
+    protected function getData(string $column, $defaultValue = null)
```
Comment (corpsee) [inline src/Model.php:47]: May be better would be `getDataItem`/`setDataItem`.

File: src/Model/Base/BuildError.php [hunk 4]
```diff
@@ -19,241 +19,120 @@ class BuildError extends Model
     use HasCreateDateTrait;
 
     public const SEVERITY_CRITICAL = 0;
-    public const SEVERITY_HIGH     = 1;
```
Comment (corpsee) [inline src/Model/Base/BuildError.php:22]: Could you revert unnecessary changes with arrays reformat, please.

File: src/Model.php (hunk 1 above)
Comment (KieranFYI) [reply src/Model.php:21]: More then happen to change it, only reason I called it `$casts` is because I mainly work with Laravel and that's what it's called there.

**Comment by corpsee** [comment]
@KieranFYI Are you finished? Can I merge PR?

**Comment by KieranFYI** [comment]
> @KieranFYI Are you finished? Can I merge PR?

For now yes, i plan to create another PR in the future with more test coverage.

The only thing i haven't finished is the constants as suggest by yourself but it seems my comment has disappeared. I was looking for an example of how you wanted it completed.

**Comment by corpsee** [comment]
@KieranFYI Ок, thanks. I want to merge PR now and may be fix constants later (It is not so importent, just a suggestion).
